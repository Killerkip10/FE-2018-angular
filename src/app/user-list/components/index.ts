export * from './user-chosen/user-chosen.component';
export * from './user-dropdown-list/user-dropdown-list.component';
export * from './user-dropdown-item/user-dropdown-item.component';
