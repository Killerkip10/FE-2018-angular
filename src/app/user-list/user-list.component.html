<section>
  <app-user-chosen [user]="(users$ | async).selected" (click)="dropDownList()"></app-user-chosen>

  <app-user-dropdown-list
    class="drop-down-list"
    (select)="dropDownList()"
    [users]="(users$ | async).users"
    *ngIf="dropDownListFlag">
    <mat-spinner *ngIf="!(users$ | async).loaded" [diameter]="18"></mat-spinner>
  </app-user-dropdown-list>
</section>
