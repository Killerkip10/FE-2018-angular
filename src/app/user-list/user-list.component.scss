section{
  box-sizing: border-box;
  border: 2px solid rgb(245, 245, 245);
  position: relative;
  app-user-dropdown-list{
    position: absolute;
    max-height: calc(100vh - 240px);
    overflow: auto;
    top: 100%;
    left: -2px;
    right: -2px;
  }
}
