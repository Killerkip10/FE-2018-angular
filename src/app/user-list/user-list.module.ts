import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';

import {UserListComponent} from './user-list.component';
import {
  UserChosenComponent,
  UserDropdownListComponent,
  UserDropdownItemComponent
} from './components';
import {
  MatCardModule,
  MatIconModule,
  MatInputModule,
  MatFormFieldModule,
  MatProgressSpinnerModule
} from '@angular/material';

@NgModule({
  imports: [
    CommonModule,

    MatCardModule,
    MatIconModule,
    MatInputModule,
    MatProgressSpinnerModule,
    MatFormFieldModule,
  ],
  declarations: [
    UserListComponent,
    UserChosenComponent,
    UserDropdownListComponent,
    UserDropdownItemComponent
  ],
  exports: [
    UserListComponent
  ]
})
export class UserListModule { }
