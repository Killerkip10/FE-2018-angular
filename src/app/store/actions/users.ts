import {Action} from '@ngrx/store';

import {User} from '../../models';

export const GET_USERS = '[Users] Get users';
export const FIND_USERS = '[Users] Find users';
export const SELECT_USER = '[Users] Select user';
export const EDIT_USER = '[Users] Edit user';
export const SUCCESS_EDIT = '[Users] Success edit';
export const SUCCESS = '[Users] Success';
export const ERROR = '[Users] Error';

export class GetUsers implements Action {
  public readonly type = GET_USERS;
}
export class FindUsers implements Action {
  public readonly type = FIND_USERS;

  constructor(public name: string) {}
}
export class SelectUser implements Action {
  public readonly type = SELECT_USER;

  constructor(public user: User) {}
}
export class EditUser implements Action {
  public readonly type = EDIT_USER;

  constructor(public user: User) {}
}
export class SuccessEdit implements Action {
  readonly type = SUCCESS_EDIT;

  constructor(public user: User) {}
}
export class Success implements Action {
  readonly type = SUCCESS;

  constructor(public users: User[]) {}
}
export class Error implements Action {
  readonly type = ERROR;

  constructor(public err: string) {}
}

export type Action = GetUsers
  | FindUsers
  | SelectUser
  | EditUser
  | SuccessEdit
  | Success
  | Error;
