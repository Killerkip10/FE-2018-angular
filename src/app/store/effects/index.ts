import {ProfileEffects} from './profile';
import {UsersEffects} from './users';

export const effects = [
  ProfileEffects,
  UsersEffects
];
