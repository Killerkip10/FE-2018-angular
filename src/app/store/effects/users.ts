import {Injectable} from '@angular/core';
import {HttpClient, HttpHeaders, HttpParams, HttpResponse} from '@angular/common/http';
import {catchError, map, mergeMap} from 'rxjs/operators';
import {of} from 'rxjs';

import {Actions, Effect, ofType} from '@ngrx/effects';
import {
  GET_USERS,
  FIND_USERS,
  FindUsers,
  Success,
  Error
} from '../actions/users';

import {User} from '../../../../server/models';
import {urlConfig} from '../../config';

const options = {
  headers: new HttpHeaders({'content-type': 'application/json'}),
  observe: 'response',
  withCredentials: true,
  params: null
};

@Injectable()
export class UsersEffects {
  constructor(
    private actions$: Actions,
    private http: HttpClient
  ) {}

  @Effect()
  loadUsers$ = this.actions$
    .pipe(
      ofType(GET_USERS),
      mergeMap(() => this.http.get<HttpResponse<User[]>>(urlConfig.getUsers, options as object)
        .pipe(
          map(resp => new Success(resp.body)),
          catchError(() => of(new Error('ERROR.BASE')))
        )
      )
    );

  @Effect()
  findUsers$ = this.actions$
    .pipe(
      ofType(FIND_USERS),
      mergeMap((action: FindUsers) => {
        const option = { ...options, params: new HttpParams().set('name', action.name)};

        return this.http.get<HttpResponse<User[]>>(urlConfig.findUsers, option as object)
          .pipe(
            map(resp => new Success(resp.body)),
            catchError(() => of(new Error('ERROR.BASE')))
          );
      })
    );
}
