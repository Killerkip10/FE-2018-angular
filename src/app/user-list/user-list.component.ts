import {Component, OnInit} from '@angular/core';
import {Observable} from 'rxjs';
import {first} from 'rxjs/operators';

import {select, Store} from '@ngrx/store';
import {getUsersJsState} from '../store/reducers/users';
import {State} from '../store/reducers/users';
import {GetUsers} from '../store/actions/users';

@Component({
  selector: 'app-user-list',
  templateUrl: './user-list.component.html',
  styleUrls: ['./user-list.component.scss']
})
export class UserListComponent implements OnInit {
  public dropDownListFlag = false;
  public users$: Observable<State>;

  constructor(private store: Store<State>) {}

  ngOnInit(): void {
    this.users$ = this.store.pipe(select(getUsersJsState));
  }

  public dropDownList(): void {
    this.dropDownListFlag = !this.dropDownListFlag;

    this.users$
      .pipe(first())
      .subscribe(state => {
        if (!state.loaded) {
          this.store.dispatch(new GetUsers());
        }
      });
  }
}
