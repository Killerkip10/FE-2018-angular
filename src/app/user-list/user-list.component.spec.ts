import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {NO_ERRORS_SCHEMA} from '@angular/core';
import {CommonModule} from '@angular/common';
import {TranslateModule} from '@ngx-translate/core';
import {of} from 'rxjs';

import {StoreModule} from '@ngrx/store';
import {EffectsModule} from '@ngrx/effects';
import {reducers} from '../store/reducers';
import {effects} from '../store/effects';
import {State} from '../store/reducers/users';

import {CoreModule} from '../core/core.module';
import {AppTranslate} from '../app.translate';
import {UserListComponent } from './user-list.component';

describe('UserListComponent', () => {
  let component: UserListComponent;
  let fixture: ComponentFixture<UserListComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        CoreModule,
        CommonModule,
        TranslateModule,

        StoreModule.forRoot(reducers),
        EffectsModule.forRoot(effects),

        AppTranslate
      ],
      declarations: [
        UserListComponent
      ],
      schemas: [
        NO_ERRORS_SCHEMA
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(UserListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('snapshot', () => {
    component.dropDownListFlag = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('snapshot', () => {
    component.dropDownListFlag = true;
    component.users$ = of({loaded: true} as State);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('snapshot', () => {
    component.users$ = of({loaded: true} as State);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('snapshot', () => {
    component.users$ = of({loaded: false} as State);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
