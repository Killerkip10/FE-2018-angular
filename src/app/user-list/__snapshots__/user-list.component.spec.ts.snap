// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListComponent snapshot 1`] = `
<app-user-list
  dropDownListFlag="false"
  store={[Function Store]}
  users$={[Function Store]}
>
  <section>
    <app-user-chosen />
    
  </section>
</app-user-list>
`;

exports[`UserListComponent snapshot 2`] = `
<app-user-list
  dropDownListFlag={[Function Boolean]}
  store={[Function Store]}
  users$={[Function Store]}
>
  <section>
    <app-user-chosen />
    
    <app-user-dropdown-list
      class="drop-down-list"
    >
      
      <mat-spinner />
    </app-user-dropdown-list>
  </section>
</app-user-list>
`;

exports[`UserListComponent snapshot 3`] = `
<app-user-list
  dropDownListFlag={[Function Boolean]}
  store={[Function Store]}
  users$={[Function Observable]}
>
  <section>
    <app-user-chosen />
    
    <app-user-dropdown-list
      class="drop-down-list"
    >
      
    </app-user-dropdown-list>
  </section>
</app-user-list>
`;

exports[`UserListComponent snapshot 4`] = `
<app-user-list
  dropDownListFlag="false"
  store={[Function Store]}
  users$={[Function Observable]}
>
  <section>
    <app-user-chosen />
    
  </section>
</app-user-list>
`;

exports[`UserListComponent snapshot 5`] = `
<app-user-list
  dropDownListFlag="false"
  store={[Function Store]}
  users$={[Function Observable]}
>
  <section>
    <app-user-chosen />
    
  </section>
</app-user-list>
`;
